src/Common/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -22,5 +26,138 @@ internal static partial class DelegateHelpers
         public delegate object VBCallSiteDelegate5<T>(T callSite, object instance, ref object arg1, ref object arg2, ref object arg3, ref object arg4, ref object arg5);
         public delegate object VBCallSiteDelegate6<T>(T callSite, object instance, ref object arg1, ref object arg2, ref object arg3, ref object arg4, ref object arg5, ref object arg6);
         public delegate object VBCallSiteDelegate7<T>(T callSite, object instance, ref object arg1, ref object arg2, ref object arg3, ref object arg4, ref object arg5, ref object arg6, ref object arg7);
+
+        internal static Delegate CreateObjectArrayDelegate(Type delegateType, System.Func<object[], object> handler)
+        {
+#if FEATURE_REFEMIT
+            return CreateObjectArrayDelegateRefEmit(delegateType, handler);
+#else
+            return Internal.Runtime.Augments.DynamicDelegateAugments.CreateObjectArrayDelegate(delegateType, handler);
+#endif
+        }
+
+
+#if FEATURE_REFEMIT
+
+        // We will generate the following code:
+        //  
+        // object ret;
+        // object[] args = new object[parameterCount];
+        // args[0] = param0;
+        // args[1] = param1;
+        //  ...
+        // try {
+        //      ret = handler.Invoke(args);
+        // } finally {
+        //      param0 = (T0)args[0];   // only generated for each byref argument
+        // }
+        // return (TRet)ret;
+        private static Delegate CreateObjectArrayDelegateRefEmit(Type delegateType, System.Func<object[], object> handler)
+        {
+            MethodInfo delegateInvokeMethod = delegateType.GetMethod("Invoke");
+
+            Type returnType = delegateInvokeMethod.ReturnType;
+            bool hasReturnValue = returnType != typeof(void);
+
+            ParameterInfo[] parameters = delegateInvokeMethod.GetParameters();
+            Type[] paramTypes = new Type[parameters.Length + 1];
+            paramTypes[0] = typeof(Func<object[], object>);
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                paramTypes[i + 1] = parameters[i].ParameterType;
+            }
+
+            DynamicMethod thunkMethod = new DynamicMethod("Thunk", returnType, paramTypes);
+            ILGenerator ilgen = thunkMethod.GetILGenerator();
+
+            LocalBuilder argArray = ilgen.DeclareLocal(typeof(object[]));
+            LocalBuilder retValue = ilgen.DeclareLocal(typeof(object));
+
+            // create the argument array
```

Comment:
Yes. In a number of places the change has been made already in one or another branch. Having  just one place for fixes like this is my priority here. I will make a note about checking if Array.Empty can be used in more places.